File: android/src/main/java/getcapacitor/community/contacts/BiMap.java
package getcapacitor.community.contacts;

import java.util.HashMap;

// Helper to enable bidirectional mapping.
public class BiMap<K, V> {

    // The following map can be used for retrieving the key belonging to a certain value.
    // This map is essentially the inverted version of `mapForValues`
    private final HashMap<V, K> mapForKeys = new HashMap<>();
    // The following map can be used for retrieving the value belonging to a certain key.
    private final HashMap<K, V> mapForValues;

    private final K defaultKey;
    private final V defaultValue;

    BiMap(HashMap<K, V> map, K defaultKey, V defaultValue) {
        this.mapForValues = map;

        for (HashMap.Entry<K, V> entry : map.entrySet()) {
            K key = entry.getKey();
            V value = entry.getValue();
            this.mapForKeys.put(value, key);
        }

        this.defaultKey = defaultKey;
        this.defaultValue = defaultValue;
    }

    K getKey(V v) {
        K k = this.mapForKeys.get(v);
        if (k == null) {
            return this.defaultKey;
        }
        return k;
    }

    V getValue(K k) {
        V v = this.mapForValues.get(k);
        if (v == null) {
            return this.defaultValue;
        }
        return v;
    }
}

File: ios/Sources/ContactsPlugin/BiMap.swift
import Foundation

// Helper to enable bidirectional mapping.
public class BiMap<K: Hashable, V: Hashable> {
    // The following map can be used for retrieving the key belonging to a certain value.
    // This map is essentially the inverted version of `mapForValues`
    private var mapForKeys: [V: K] = [:]
    // The following map can be used for retrieving the value belonging to a certain key.
    private var mapForValues: [K: V] = [:]

    private var defaultKey: K
    private var defaultValue: V

    init(_ map: [K: V], _ defaultKey: K, _ defaultValue: V) {
        self.mapForValues = map

        for entry in map {
            let key: K = entry.key
            let value: V = entry.value
            self.mapForKeys[value] = key
        }

        self.defaultKey = defaultKey
        self.defaultValue = defaultValue
    }

    func getKey(_ v: V?) -> K {
        let k: K? = self.mapForKeys[v ?? "" as! V]
        guard let k = k else {
            return self.defaultKey
        }
        return k
    }

    func getValue(_ k: K?) -> V {
        let v: V? = self.mapForValues[k ?? "" as! K]
        guard let v = v else {
            return self.defaultValue
        }
        return v
    }
}
